stage1/init/kvm.go

```diff
@@ -113,16 +115,39 @@ func mountSharedVolumes(root string, p *stage1commontypes.Pod, ra *schema.Runtim
 }
 
 func doBindMount(source, destination string, readOnly bool, recursive *bool) error {
-	var flags uintptr = syscall.MS_BIND | syscall.MS_REC
+	var flags uintptr = syscall.MS_BIND
+
 	if readOnly {
 		flags |= syscall.MS_RDONLY
 	}
 
 	// Enable recursive by default and remove it if explicitly requested
-	if recursive != nil && *recursive == false {
-		flags ^= syscall.MS_REC
+	recursiveBool := recursive == nil || *recursive == true
+	if recursiveBool {
+		flags |= syscall.MS_REC
+	}
+
+	if err := syscall.Mount(source, destination, "bind", flags, ""); err != nil {
+		return err
 	}
-	return syscall.Mount(source, destination, "bind", flags, "")
+
+	if readOnly && recursiveBool {
+		// Sub-mounts are still read-write, so find them and remount them read-only
+
+		mnts, err := stage0.GetMountsForPrefix(source + "/")
+		if err != nil {
+			return errwrap.Wrap(fmt.Errorf("error getting mounts under %q from mountinfo", source), err)
+		}
+
+		for _, mnt := range mnts {
+			innerAbsPath := destination + strings.Replace(mnt.MountPoint, source, "", -1)
+			if err := syscall.Mount("", innerAbsPath, "none", syscall.MS_REMOUNT|syscall.MS_RDONLY, ""); err != nil {
+				return err
```

Comment:
Mounts listed in mountinfo might not all be reachable: some might be masked mounts.  In this case, mount will return ENOENT or EINVAL. The error in this case should be ignored. Alternatively, we can detect if `innerAbsPath` is a mount point. See https://github.com/systemd/systemd/issues/3793 and https://github.com/systemd/systemd/pull/3794 for the details.

/cc @iaguis do we have a function in Go to detect mountpoints?